encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/TypeToken.java

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders;
+
+import androidx.annotation.NonNull;
+
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.WildcardType;
+
+public abstract class TypeToken<T> {
```

Comment:
We would want to forbid any subclasses of this class other than `ClassToken` and `ArrayToken` effectively making this class "sealed". The only way to achieve it in java is to `static`ally nest `ClassToken` and `ArrayToken` inside `TypeToken` and make its constructor private.

Reply:
afaict this still needs a `private` constructor, to forbid other subclasses